File: solutions/leetcode/p76_minimum_window_substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char,int> cnt_t;
        for(const char& c : t) {
            cnt_t[c]++;
        }
        int missing = (int)cnt_t.size();

        unordered_map<char,int> cnt_s;
        int i=0, j=0, l=0, r=1e9;
        while(i<(int)s.length()){
            cnt_s[s[i]]++;
            if(cnt_s[s[i]] == cnt_t[s[i]]) missing--;
            while(!missing) {
                if(i-j < r-l){
                    l = j;
                    r = i;
                }
                if(cnt_s[s[j]] == cnt_t[s[j]]) missing++;
                cnt_s[s[j]]--;
                j++;
            }
            i++;
        }
        return (r==1e9) ? "" : s.substr(l, r-l+1);
    }
};

// expand window until t included, then shorten window as much as possible while t still in s
// can just keep l and r ptrs and use substr(l, r-l+1) at end
// can do map.size() to get size of maps
// simple sliding window problem, expand window until requirements meet, then decrease window as much as possible while still meeting requirements, keep track of best ans throughout
// changing to unordered_map greatly decreased runtime

File: solutions/leetcode/p76_minimum_window_substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        cnt = Counter(t)
        missing = len(cnt)

        i = 0
        l = 0
        r = float('inf')
        for j in range(len(s)):
            if s[j] in cnt:
                cnt[s[j]] -= 1
                if cnt[s[j]] == 0:
                    missing -= 1
            
            while missing == 0:
                if j-i < r-l:
                    r = j
                    l = i
                if s[i] in cnt:
                    cnt[s[i]] += 1
                    if cnt[s[i]] > 0:
                        missing += 1
                i += 1

        return s[l:r+1] if r != float('inf') else ""
